Gradcheck forward AD respects requires grad but run with requires_grad=False

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #72309

Fixes: https://github.com/pytorch/pytorch/issues/72113

~Hmm, currently failing for forward over reverse grad according to CI, will investigate more.~

Differential Revision: [D33991570](https://our.internmc.facebook.com/intern/diff/D33991570)

soulitzer committed:
Gradcheck forward AD respects forward AD but run with requires_grad=False

[ghstack-poisoned]

soulitzer committed:
Update on "Gradcheck forward AD respects forward AD but run with requires_grad=False"




[ghstack-poisoned]

soulitzer committed:
Update on "Gradcheck forward AD respects requires grad but run with requires_grad=False"


Fixes: https://github.com/pytorch/pytorch/issues/72113

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33991570).

albanD reviewed (commented):
You definitely want to enable grad in https://github.com/pytorch/pytorch/blob/1e7d20eaea2071661545be5b6ca54eda15c92fe3/torch/autograd/gradcheck.py#L1539-L1542 ;)

soulitzer committed:
Update on "Gradcheck forward AD respects requires grad but run with requires_grad=False"


Fixes: https://github.com/pytorch/pytorch/issues/72113

Hmm, currently failing for forward over reverse grad according to CI, will investigate more.

Differential Revision: [D33991570](https://our.internmc.facebook.com/intern/diff/D33991570)

[ghstack-poisoned]

soulitzer committed:
Update on "Gradcheck forward AD respects requires grad but run with requires_grad=False"


Fixes: https://github.com/pytorch/pytorch/issues/72113

Hmm, currently failing for forward over reverse grad according to CI, will investigate more.

Differential Revision: [D33991570](https://our.internmc.facebook.com/intern/diff/D33991570)

[ghstack-poisoned]

soulitzer commented:
> You definitely want to enable grad in 

Good catch!

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33991570).

albanD commented on test/test_autograd.py:
```diff
@@ -4064,6 +4064,60 @@ def basic_mul(x):
                 gradcheck(bad_fn, (x, y), check_forward_ad=True, fast_mode=fast_mode)
 
+    def test_gradcheck_forward_ad_respects_requires_grad_but_runs_with_no_requires_grad(self):
+        # Currently requires_grad is used as a easy way for gradcheck to know
```
Not sure what this test is checking? That the requires_grad flag is properly respected by the forward ad gradcheck?

soulitzer replied:
Hmm maybe I can improve the comments in this test here,
but yeah we want to make sure 1) requires_grad is properly respected by the various checks in gradcheck, and 2) that the actual test is run with requires_grad=False

for (1), we make sure the jvp count is what we expect, and for (2) we look at vjp to make sure it is 0 when check_backward_ad=False

albanD replied:
> (2) we look at vjp to make sure it is 0 when check_backward_ad=False

That only means that no backward() call happened. Which has always been True. It does not check that we don't run the forward with requires_grad=False right?
You could add a check in the function given to gradcheck to ensure that?

albanD commented on torch/autograd/gradcheck.py:
```diff
@@ -1531,14 +1534,15 @@ def randn_like(x):
-        input_args = tuple(x for x in input_args
-                           if is_tensor_like(x) and x.requires_grad)
-        grad_inputs = torch.autograd.grad(outputs, input_args, grad_outputs, create_graph=True,
+        grad_outputs = tuple(x.requires_grad_() for x in args[-num_outputs:])
```
nit: add a comment saying that all grad_outputs always require gradients in the gradgradcheck context?

soulitzer replied:
Found out from a test failure that is not actually the case! for certain ops the grad_outputs don't necessarily require grad. I've updated the code.

albanD commented on torch/autograd/gradcheck.py:
```diff
@@ -1531,14 +1534,15 @@ def randn_like(x):
-                           if is_tensor_like(x) and x.requires_grad)
-        grad_inputs = torch.autograd.grad(outputs, input_args, grad_outputs, create_graph=True,
+        grad_outputs = tuple(x.requires_grad_() for x in args[-num_outputs:])
+        diff_input_args = tuple(x for i, x in enumerate(input_args) if i in diff_input_args_indices)
```
You just fixed the requires_grad fields, so no need to change this?

soulitzer replied:
yeah there should be no need, but this one I prefer because we can avoid repeating checking is_tensor_like and requires grad again. diff_input_args also better reflects what the tuple contains compared to input_args

soulitzer committed:
Update on "Gradcheck forward AD respects requires grad but run with requires_grad=False"


Fixes: https://github.com/pytorch/pytorch/issues/72113

~Hmm, currently failing for forward over reverse grad according to CI, will investigate more.~

Differential Revision: [D33991570](https://our.internmc.facebook.com/intern/diff/D33991570)

[ghstack-poisoned]

soulitzer committed:
Update on "Gradcheck forward AD respects requires grad but run with requires_grad=False"


Fixes: https://github.com/pytorch/pytorch/issues/72113

~Hmm, currently failing for forward over reverse grad according to CI, will investigate more.~

Differential Revision: [D33991570](https://our.internmc.facebook.com/intern/diff/D33991570)

[ghstack-poisoned]

albanD reviewed (approved):
Minor test update, SGTM otherwise.

soulitzer committed:
Update on "Gradcheck forward AD respects requires grad but run with requires_grad=False"


Fixes: https://github.com/pytorch/pytorch/issues/72113

~Hmm, currently failing for forward over reverse grad according to CI, will investigate more.~

Differential Revision: [D33991570](https://our.internmc.facebook.com/intern/diff/D33991570)

[ghstack-poisoned]

soulitzer committed:
Update on "Gradcheck forward AD respects requires grad but run with requires_grad=False"


Fixes: https://github.com/pytorch/pytorch/issues/72113

~Hmm, currently failing for forward over reverse grad according to CI, will investigate more.~

Differential Revision: [D33991570](https://our.internmc.facebook.com/intern/diff/D33991570)

[ghstack-poisoned]

soulitzer committed:
Update on "Gradcheck forward AD respects requires grad but run with requires_grad=False"


Fixes: https://github.com/pytorch/pytorch/issues/72113

~Hmm, currently failing for forward over reverse grad according to CI, will investigate more.~

Differential Revision: [D33991570](https://our.internmc.facebook.com/intern/diff/D33991570)

[ghstack-poisoned]

soulitzer commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -3063,10 +3068,7 @@ Tensor& linalg_svdvals_out(const Tensor& A, Tensor & S) {
 Tensor linalg_svdvals(const Tensor& A) {
-  const bool A_requires_grad = (at::GradMode::is_enabled() && A.requires_grad())
-                               || A._fw_grad(/*level */ 0).defined()
-                               || isTensorSubclassLike(A);
```
@Lezcano functorch still seems to pass without `isTensorSubclassLike` here

Lezcano replied:
Hmmm. Some tests from functorch (something like "batched over jvp/vjp" or so) did not pass a week ago when we tested this. This was because the wrapper tensor did not have `requires_grad=True` nor a forward grad. What changed that makes it work now?

soulitzer replied:
That is weird... here are some more data points:

Without `isTensorSubclassLike`:
- Without the forward grad defined check I run into `RuntimeError: U_.dim() >= 2 && Vh_.dim() >= 2INTERNAL ASSERT FAILED at "../torch/csrc/autograd/FunctionsManual.cpp":2391, please report a bug to PyTorch.` - for `test_jvp`, `test_vmapjvp` and `test_vmapjvpall` (linalg_svdvals_cpu_float32).
- With the forward grad defined check, all 3 tests pass.

With `isTensorSubclassLike`:
- all 3 pass with or without forward grad defined check

To repro:
- run something like `python test/test_ops.py -k test_jvp_linalg_svdvals`

soulitzer replied:
This PR only changes the testing side, but with the updated testing I found that `eigvals` and `maxpool_1d` needed similar changes to get the tests to pass. `eigvalsh` seemed like a similar case, so I also updated that.

Lezcano replied:
I mean, I'm happy if the tests pass. We just added that condition to make some tests pass, but if those are green I'm happy for this condition to go.

But again, it's odd, as when we debugged this, we saw that the external tensor did not have `requires_grad=True`, so it was reasonable for the test to fail. Let's just hope that it does not fail later on, and if it does, that `INTERNAL_ASSERT` will fail and we'll know what to revert :P

soulitzer commented on aten/src/ATen/native/EmbeddingBag.cpp:
```diff
@@ -798,7 +798,7 @@ embedding_bag(const Tensor &weight, const Tensor &indices,
     padding_idx = maybe_wrap_dim(padding_idx, weight.size(0));
   }
   std::tuple<Tensor, Tensor, Tensor, Tensor> out;
-  if (!weight.requires_grad()) {
```
@zou3519 these might fix some failures on the functorch side

see https://github.com/pytorch/functorch/pull/473

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33991570).

albanD reviewed (approved):
Nice fix!
